docs/decisions/0004-lti-advantage-nrps.rst

```diff
@@ -0,0 +1,84 @@
+
+From the LTI-NRPS Specification:
+
+    The Names and Role Provisioning Services is based on IMS Learning Information Services (LIS) [LIS-20]
+    and W3C Organization Ontology [W3C-ORG]. It is concerned with providing access to data about users’ roles
+    within organizations, a course being an example of an organization. So a very common purpose for this service
+    is to provide a roster (list of enrolments) for a course.
+
+The LTI-NRPS services has two services which provide user data in the following format:
+
+.. code-block:
+    {
+        "id" : "https://lms.example.com/sections/2923/memberships",
+        "context": {
+            "id": "2923-abc",
+            "label": "CPS 435",
+            "title": "CPS 435 Learning Analytics",
+        },
+        "members" : [
+            {
+                "status" : "Active",
+                "name": "Jane Q. Public",
+                "picture" : "https://platform.example.edu/jane.jpg",
+                "given_name" : "Jane",
+                "family_name" : "Doe",
+                "middle_name" : "Marie",
+                "email": "jane@platform.example.edu",
+                "user_id" : "0ae836b9-7fc9-4060-006f-27b2066ac545",
+                "lis_person_sourcedid": "59254-6782-12ab",
+                "roles": [
+                    "http://purl.imsglobal.org/vocab/lis/v2/membership#Instructor"
+                ]
+            }
+        ]
+    }
+
+Where the only required fields for each member are :code:`user_id` and :code:`roles`, to avoid unintended PII transmission to a third party.
+But since one of the purposes of this API is to enable tools access to user PII, the implementation should provide a flag to enable the consumer
+to pass additional PII parameters such as name, email, and a profile picture url.
+
+The main issue with enabling this API (`even with pagination as defined in the spec`_) is that it would allow tools to scrape enrollments for all
+students in a given course. Given the scale of some courses in edX.org, even if the API is made more performant with cursor pagination,
+we're still left with an API pattern that would be actively encouraging clients out there to make hundreds of thousands of successive requests in order
+to crawl enrollment data.
+
+.. _`Names and Roles Provisioning Services (NRPS)`: http://www.imsglobal.org/spec/lti-nrps/v2p0
+.. _`even with pagination as defined in the spec`: http://www.imsglobal.org/spec/lti-nrps/v2p0#limit-query-parameter
+
+Proposed solutions
+==================
+
+Don't implement NRPS
+~~~~~~~~~~~~~~~~~~~~
+Placeholder.
+
+Implement NRPS gated by course
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Placeholder.
+
+Implement NRPS for courses up to a predefined number of students
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Placeholder.
+
+Implement NRPS limiting the context of the data retrieved
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Placeholder.
+
+Consequences
+============
```

Comment:
Another option is to pre-calculate the enrollment data (say on a nightly basis) and make a giant file that we push up to S3 and redirect to with an authorized link. That would allow us to serve the data quickly, though it would mean that the data was stale. Also, there's a good chance that getting a single file with a million enrollments would choke any potential consumer... but it is an option.

Reply:
We can actually pre-calculate multiple files in a scheduled task. And use those files as paginations. That will not decrease the number of requests, but requests will not do any db queries.